internal/controllers/machine/machine_controller_phases_test.go

```diff
@@ -1003,82 +1080,170 @@ func TestReconcileInfrastructure(t *testing.T) {
-						ConfigRef: &corev1.ObjectReference{
-							APIVersion: "bootstrap.cluster.x-k8s.io/v1beta1",
-							Kind:       "GenericBootstrapConfig",
-							Name:       "bootstrap-config1",
-						},
-					},
 					InfrastructureRef: corev1.ObjectReference{
 						APIVersion: "infrastructure.cluster.x-k8s.io/v1beta1",
 						Kind:       "GenericInfrastructureMachine",
 						Name:       "infra-config1",
 					},
+					ProviderID:    ptr.To("test://something"),
+					FailureDomain: ptr.To("something"),
 				},
 				Status: clusterv1.MachineStatus{
-					BootstrapReady:      true,
 					InfrastructureReady: true,
-					NodeRef:             &corev1.ObjectReference{Kind: "Node", Name: "machine-test-node"},
+					Addresses: []clusterv1.MachineAddress{
+						{
+							Type:    clusterv1.MachineExternalIP,
+							Address: "1.2.3.4",
+						},
+					},
 				},
 			},
-			bootstrapConfig: map[string]interface{}{
-				"kind":       "GenericBootstrapConfig",
-				"apiVersion": "bootstrap.cluster.x-k8s.io/v1beta1",
+			infraMachine: map[string]interface{}{
+				"kind":       "GenericInfrastructureMachine",
+				"apiVersion": "infrastructure.cluster.x-k8s.io/v1beta1",
 				"metadata": map[string]interface{}{
-					"name":      "bootstrap-config1",
+					"name":      "infra-config1",
 					"namespace": metav1.NamespaceDefault,
 				},
-				"spec": map[string]interface{}{},
+				"spec": map[string]interface{}{
+					"providerID":    "test://id-1",
+					"failureDomain": "foo",
+				},
 				"status": map[string]interface{}{
-					"ready":          true,
-					"dataSecretName": "secret-data",
+					"ready": false,
+					"addresses": []interface{}{
+						map[string]interface{}{
+							"type":    "InternalIP",
+							"address": "10.0.0.1",
+						},
+						map[string]interface{}{
+							"type":    "InternalIP",
+							"address": "10.0.0.2",
+						},
+					},
 				},
 			},
-			infraConfig: map[string]interface{}{
-				"kind":       "GenericInfrastructureMachine",
-				"apiVersion": "infrastructure.cluster.x-k8s.io/v1beta1",
-				"metadata":   map[string]interface{}{},
-			},
-			expectResult: ctrl.Result{},
-			expectError:  true,
+			InfraMachineGetError: nil,
+			expectResult:         ctrl.Result{},
+			expectError:          false,
 			expected: func(g *WithT, m *clusterv1.Machine) {
-				g.Expect(m.Status.InfrastructureReady).To(BeTrue())
-				g.Expect(m.Status.FailureMessage).NotTo(BeNil())
-				g.Expect(m.Status.FailureReason).NotTo(BeNil())
-				g.Expect(m.Status.GetTypedPhase()).To(Equal(clusterv1.MachinePhaseFailed))
+				g.Expect(m.Status.InfrastructureReady).To(BeFalse())
+				g.Expect(ptr.Deref(m.Spec.ProviderID, "")).To(Equal("test://something"))
+				g.Expect(ptr.Deref(m.Spec.FailureDomain, "")).To(Equal("something"))
+				g.Expect(m.Status.Addresses).To(HaveLen(1))
 			},
 		},
 		{
-			name: "infrastructure ref is paused",
-			infraConfig: map[string]interface{}{
+			name: "should change data also after infrastructure ready is set",
```

Comment:
I'm not entirely sure it is correct to allow changing all the data after infrastructure ready,
E.g. is it correct to allow changing the ProviderID? (I have less concerns about failureDomain and addresses)

Reply:
Hmm this might depend, let's leave it as is?

Reply:
ACK, let's tackle this separately